src/collection-view.js

```diff
@@ -46,33 +50,42 @@ Marionette.CollectionView = Marionette.View.extend({
   },
 
   endBuffering: function() {
+    // Only trigger attach if already shown and attached, otherwise Region#show() handles this.
+    var canTriggerAttach = this._isShown && Marionette.isNodeAttached(this.el);
+    var nestedViews;
+
     this.isBuffering = false;
-    this._triggerBeforeShowBufferedChildren();
 
-    this.attachBuffer(this);
+    if (this._isShown) {
+      this._triggerMethodOnAll(this._bufferedChildren, 'before:show');
+    }
+    if (canTriggerAttach && this.triggerBeforeAttach) {
+      // Recursively retrieve nested views.
+      nestedViews = this._getNestedViews();
+      this._triggerMethodOnAll(nestedViews, 'before:attach');
+    }
 
-    this._triggerShowBufferedChildren();
-    this.initRenderBuffer();
-  },
+    this.attachBuffer(this);
 
-  _triggerBeforeShowBufferedChildren: function() {
+    if (canTriggerAttach && this.triggerAttach) {
+      // Don't recursively retrieve nested views unnecessarily.
+      nestedViews = nestedViews || this._getNestedViews();
+      this._triggerMethodOnAll(nestedViews, 'attach');
+    }
     if (this._isShown) {
-      _.each(this._bufferedChildren, _.partial(this._triggerMethodOnChild, 'before:show'));
+      this._triggerMethodOnAll(this._bufferedChildren, 'show');
     }
-  },
 
-  _triggerShowBufferedChildren: function() {
-    if (this._isShown) {
-      _.each(this._bufferedChildren, _.partial(this._triggerMethodOnChild, 'show'));
+    this.initRenderBuffer();
+  },
 
-      this.initRenderBuffer();
-    }
+  _triggerMethodOnAll: function(views, event) {
+    _.each(views, _.partial(this._triggerMethodOn, event));
   },
 
-  // Internal method for _.each loops to call `Marionette.triggerMethodOn` on
-  // a child view
-  _triggerMethodOnChild: function(event, childView) {
-    Marionette.triggerMethodOn(childView, event);
+  // Internal method for _.each loops to call `Marionette.triggerMethodOn` on a view
+  _triggerMethodOn: function(event, view) {
```

Comment:
I was wondering about that.  Delegate straight to Marionette.triggerMethodOn, then?  Or you mean leave `_triggerMethodOnChild` alone?